statsmodels/regression/heckman.py

```diff
@@ -0,0 +1,655 @@
+
+        ## Show warning to user if estimation is by two-step but MLE arguments were also provided
+        if method=='twostep':
+            if start_params_mle is not None or method_mle is not None or maxiter_mle is not None or \
+                len(kwargs_mle.keys())>0:
+                warnings.warn('The user chose to estimate the Heckman model by Two-Step,' + \
+                    ' but MLE arguments were provided. Extraneous MLE arguments will be ignored.')
+
+
+        ## fit
+        if method=='twostep':
+            results = self._fit_twostep()
+        elif method=='mle':
+            results = self._fit_mle(
+                start_params_mle=start_params_mle, method_mle=method_mle, maxiter_mle=maxiter_mle, 
+                **kwargs_mle)
+        else:
+            raise ValueError("Invalid choice for estimation method.")
+
+        
+        ## return fitted Heckman model
+        return results
+
+
+    def _fit_twostep(self):
+        ########################################################################
+        # PRIVATE METHOD
+        # Fits using Heckman two-step from Heckman (1979).
+        ########################################################################
+
+        ## prep data
+        Y, X, Z = self.get_datamats()
+
+        ## Step 1
+        step1model = sm.Probit(self.treated, Z)
+        step1res = step1model.fit(disp=False)
+        step1_fitted = np.atleast_2d(step1res.fittedvalues).T
+        step1_varcov = step1res.cov_params()
+
+        inverse_mills = norm.pdf(step1_fitted)/norm.cdf(step1_fitted)
+
+        ## Step 2
+        W = np.hstack((X, inverse_mills[self.treated] ) )
+        step2model = sm.OLS(Y, W)
+        step2res = step2model.fit()
+
+        params = step2res.params[:-1]
+        betaHat_inverse_mills = step2res.params[-1]
+
+
+        ## Compute standard errors
+        # Compute estimated error variance of censored regression
+        delta = np.multiply(inverse_mills, inverse_mills + step1_fitted)[self.treated]
+
+        sigma2Hat = step2res.resid.dot(step2res.resid) / self.nobs_uncensored + \
+            (betaHat_inverse_mills**2 * sum(delta)) / self.nobs_uncensored
+        sigma2Hat = sigma2Hat[0]
+        sigmaHat = np.sqrt(sigma2Hat)
+        rhoHat = betaHat_inverse_mills / sigmaHat
+
+        # compute standard errors of beta estimates of censored regression
+        R = np.zeros([self.nobs_uncensored, self.nobs_uncensored])
+        for i in range(self.nobs_uncensored):
+            R[i,i] = 1 - rhoHat**2 * delta[i]
+
+        D = np.zeros([self.nobs_uncensored, self.nobs_uncensored])
+        for i in range(self.nobs_uncensored):
+            D[i,i] = delta[i]
+
+
+        
+        Q = rhoHat**2 * (W.T.dot(D).dot(Z[self.treated])).dot(step1_varcov).dot(Z[self.treated].T.dot(D).dot(W))
+        
+        normalized_varcov_all = np.linalg.inv(W.T.dot(W)).dot(W.T.dot(R).dot(W)+Q).dot(np.linalg.inv(W.T.dot(W)))
+        normalized_varcov = normalized_varcov_all[:-1,:-1]
```

Comment:
Create temp variables to avoid duplicate calculations 
E.g. np.linalg.inv(W.T.dot(W))